### PR TITLE
Share cache for different object file names with -MD/-MMD

### PR DESCRIPTION
**Pull request by dyadav7**
This is fix for #524.

### CODE REVIEW

**Comment by dyadav7** [comment]
I will take care of regression and re submit another pull request

**Comment by dyadav7** [comment]
i closed the pull request accidentally, Now  re-opening it

File: doc/MANUAL.adoc [hunk 1]
```diff
@@ -1220,6 +1220,11 @@ filesystems), but keep in mind that:
 * Having the cache on NFS may slow down compilation. Make sure to do some
   benchmarking to see if it's worth it.
 * ccache hasn't been tested very thoroughly on NFS.
+* If operating system versions is different then cache direct hit will be lesser,
+  and it will lead to slower cache preprocessor hits. sloppiness config can be set to 
+  system_headers to ignore different system header and increase cache direct hits.
+  It is recommended to have same operating system version when using cache on NFS
+  to increase cache direct hits.
```
[the remark this answers is not in the document]
Comment (dyadav7) [reply doc/MANUAL.adoc]: i am new to github. i i have now filed a separate PR #604 for documentation changes.

File: src/argprocessing.cpp [hunk 2]
```diff
@@ -907,6 +908,13 @@ process_args(Context& ctx,
     }
   }
 
+  if (!state.dependency_target_specified
+      && ctx.args_info.generating_dependencies
+      && ctx.args_info.seen_MD_MMD)
```
[the remark this answers is not in the document]
Comment (dyadav7) [reply src/argprocessing.cpp]: done

File: src/ArgsInfo.hpp [hunk 3]
```diff
@@ -58,6 +58,12 @@ struct ArgsInfo
   // Is the compiler being asked to output dependencies?
   bool generating_dependencies = false;
 
+  // Option MD/MMD
+  bool seen_MD_MMD = false;
+
+  // Change target of dependency file
+  bool change_dep_file = false;
```
[the remark this answers is not in the document]
Comment (dyadav7) [reply src/ArgsInfo.hpp:65]: change_dep_file depends on `ArgumentProcessingState `and `ArgsInfo`. So, i saved the flag in `ArgsInfo`.
```
  if (!state.dependency_target_specified && ctx.args_info.seen_MD_MMD)
  {
    ctx.args_info.change_dep_file = true;
  }
```
The `state` variable is local to function `process_args`. How to pass `ArgumentProcessingState `to `bool change_dep_file()` in later calls . ?

[the remark this answers is not in the document]
Comment (jrosdahl) [reply src/ArgsInfo.hpp:65]: > The `state` variable is local to function `process_args`. How to pass `ArgumentProcessingState `to `bool change_dep_file()` in later calls . ?

See _"Therefore I would prefer to store `dependency_target_specified` in `ArgsInfo`"_ in my previous comment.

File: src/result.hpp [hunk 4]
```diff
@@ -44,7 +44,29 @@ enum class FileType : UnderlyingFileTypeInt {
   dwarf_object = 6,
 };
 
-using ResultFileMap = std::map<FileType, std::string /*path*/>;
+class WriteFd {
```
[the remark this answers is not in the document]
Comment (jrosdahl) [reply src/result.hpp:47]: I have now improved the result API and code, see ee7a1b0c2584. Modification of the dependency file data can now be done in `ResultRetriever::on_entry_data` if `m_dest_file_type == FileType::dependency`.

File: src/ArgsInfo.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dyadav7) [reply src/ArgsInfo.hpp:65]: Moved `dependency_target_specified `in `ArgsInfo`

File: src/result.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dyadav7) [reply src/result.hpp]: now used the `ResultRetriever::on_entry_data` .

File: src/ArgsInfo.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrosdahl) [reply src/ArgsInfo.hpp:65]: > now used the `ResultRetriever::on_entry_data` .

OK, but `change_dep_file` is still left in `ArgsInfo`? Quoting my earlier reply: _"[...] but `change_dep_file` is a decision based on the collected facts, not a fact by itself. Therefore I would prefer to store `dependency_target_specified` in `ArgsInfo` and **add a `bool change_dep_file(const ArgsInfo&)` function** in `ccache.cpp` instead. Sounds reasonable?"_. In other words, I don't think that `change_deps_file` belongs to `ArgsInfo`.

[the remark this answers is not in the document]
Comment (jrosdahl) [reply src/ArgsInfo.hpp:65]: I'll fix this after merging.

**Pull request merged**